### PR TITLE
Fix atomicity in SQLiteClusterMessenger when SQLiteNode doesn't exist

### PR DESCRIPTION
**Pull request by tylerkaraszewski**
### Details
This cleans up atomicity in `SQLiteClusterManager`. Rather than taking a *reference* to a `shared_ptr` to a `SQLiteNode`, we now take a `shared_ptr` directly (and it is `const` as well). Because it is not a reference, and made clear by being `const`, once this `SQLiteClusterManager` is created, the `SQLiteNode` it's pointer refers to will not change. This means that we do not need to check for the validity of this pointer. If it was ever valid, it is still valid (and we only ever create it with a valid object).

However, this `const shared_ptr` necessitates changing the lifetime of the `SQLiteClusterManager`, which now can't be created before our sync node is created, as we need a `shared_ptr` to it. We now create an `SQLiteClusterManager` immediately after creating the sync node, and release it immediately before destroying the sync node.

This in turn, necessitates changes to our threads that access `SQLiteClusterManager`. Because the object is now created and destroyed with the sync thread, we make a `shared_ptr` to it in `BedrockServer` so that the other threads besides the sync thread (where it is created and destroyed) can access it through the pointer (which is null when the sync thread is not running).

The sync thread sets and clears this `shared_ptr`. Every thread that wants to access the `SQLiteClusterManager` now does so by copying the `shared_ptr` to it. Once the copy is made, if it contains a value, that `SQLiteClusterManager` is guaranteed to stay in scope until the copy is destroyed. Because the `SQLiteClusterManager` in turn contains a `shared_ptr` to the sync node, that object is also guaranteed to stay in scope until all copies of the `SQLiteClusterManager` have been destroyed.

Additionally, this change `_shutDownBy` in `SQLiteClusterManager` to have lock-free atomicity by using an `atomic_flag` to indicate if a shutdown time has been set. A shutdown flag can only be set once, and to be re-used, you need to re-create a new `SQLiteClusterManager`, but this is how `BedrockServer` manages this object's lifetime now anyway.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/210104

### Tests
All FuzzyBot tests pass ([256a3da](https://github.com/Expensify/Bedrock/pull/1270/commits/256a3da4b92ad5d8bd80d079a59afff0a4c4c988))
All EBM Tests pass ([256a3da](https://github.com/Expensify/Bedrock/pull/1270/commits/256a3da4b92ad5d8bd80d079a59afff0a4c4c988))
All Auth tests pass ([256a3da](https://github.com/Expensify/Bedrock/pull/1270/commits/256a3da4b92ad5d8bd80d079a59afff0a4c4c988))

Note: Travis is only failing for internal Travis BS. Will re-run tomorrow.

### CODE REVIEW

File: BedrockServer.cpp [hunk 1]
```diff
@@ -1521,7 +1527,10 @@ void BedrockServer::postPoll(fd_map& fdm, uint64_t& nextActivity) {
     if (_shutdownState.load() == START_SHUTDOWN) {
         if (!_lastChance) {
             _lastChance = STimeNow() + 5 * 1'000'000; // 5 seconds from now.
```
Comment (rafecolton) [inline BedrockServer.cpp:1529]: Now that we're using `_shutdownSet`, do we still need to check `_lastChance` here?

**Review by rafecolton — CHANGES_REQUESTED** [review]
I think this makes `_lastChance` unnecessary? If I have that right, you can feel free to close https://github.com/Expensify/Bedrock/pull/1268 and incorporate those changes here.

File: BedrockServer.cpp [hunk 2]
```diff
@@ -704,6 +706,9 @@ void BedrockServer::sync()
         _blockingCommandQueue.clear();
     }
 
+    // We clear this before the _syncNode that it references.
+    _clusterMessenger.reset();
+
```
[the remark this answers is not in the document]
Comment (tylerkaraszewski) [reply BedrockServer.cpp:711]: https://en.cppreference.com/w/cpp/memory/shared_ptr/reset

**Comment by tylerkaraszewski** [comment]
> I think this makes `_lastChance` unnecessary? If I have that right, you can feel free to close #1268 and incorporate those changes here.

I thought we could merge this into that branch and then merge that branch to remove it.

File: BedrockServer.cpp (hunk 1 above)
Comment (tylerkaraszewski) [reply BedrockServer.cpp:1529]: Not really but I think the other PR can remove it.

**Review by rafecolton — APPROVED** [review]

**Pull request merged**